src/webgpu/shader/validation/expression/call/builtin/insertBits.spec.ts

```diff
@@ -0,0 +1,212 @@
+const builtin = 'insertBits';
+export const description = `
+Validation tests for the ${builtin}() builtin.
+`;
+
+import { makeTestGroup } from '../../../../../../common/framework/test_group.js';
+import { keysOf, objectsToRecord } from '../../../../../../common/util/data_tables.js';
+import { range } from '../../../../../../common/util/util.js';
+import {
+  Type,
+  kConcreteIntegerScalarsAndVectors,
+  kFloatScalarsAndVectors,
+  u32,
+} from '../../../../../util/conversion.js';
+import { linearRange } from '../../../../../util/math.js';
+import { ShaderValidationTest } from '../../../shader_validation_test.js';
+
+import {
+  fullRangeForType,
+  kConstantAndOverrideStages,
+  stageSupportsType,
+  validateConstOrOverrideBuiltinEval,
+} from './const_override_validation.js';
+
+export const g = makeTestGroup(ShaderValidationTest);
+
+const kValuesTypes = objectsToRecord(kConcreteIntegerScalarsAndVectors);
+
+g.test('values')
+  .desc(
+    `
+Validates that constant evaluation and override evaluation of ${builtin}() never errors on valid inputs
+`
+  )
+  .params(u =>
+    u
+      .combine('stage', kConstantAndOverrideStages)
+      .combine('type', keysOf(kValuesTypes))
+      .filter(u => stageSupportsType(u.stage, kValuesTypes[u.type]))
+      .beginSubcases()
+      .expand('value', u => fullRangeForType(kValuesTypes[u.type]))
+      .expand('newbits', u => fullRangeForType(kValuesTypes[u.type]))
+      .expand('offset', _ => linearRange(0, 32, 1).map(f => Math.floor(f)))
+      .expand('count', u => linearRange(0, 32 - u.offset, 1).map(f => Math.floor(f)))
+  )
+  .fn(t => {
+    const expectedResult = true; // insertBits() should never error
+    validateConstOrOverrideBuiltinEval(
+      t,
+      builtin,
+      expectedResult,
+      [
+        kValuesTypes[t.params.type].create(t.params.value),
+        kValuesTypes[t.params.type].create(t.params.newbits),
+        u32(t.params.offset),
+        u32(t.params.count),
+      ],
+      t.params.stage
+    );
+  });
+
+g.test('mismatched')
+  .desc(
+    `
+Validates that even with valid types, if arg0 and arg1 do not match types ${builtin}() errors
+`
+  )
+  .params(u => u.combine('arg0', keysOf(kValuesTypes)).combine('arg1', keysOf(kValuesTypes)))
+  .fn(t => {
+    const arg0 = kValuesTypes[t.params.arg0];
+    const arg1 = kValuesTypes[t.params.arg1];
+    validateConstOrOverrideBuiltinEval(
+      t,
+      builtin,
+      /* expectedResult */ t.params.arg0 === t.params.arg1,
+      [arg0.create(0), arg1.create(0), u32(0), u32(32)],
```

Comment:
```suggestion
      [arg0.create(0), arg1.create(1), u32(0), u32(32)],
```

Maybe insert a `1` just in case there is special handling for inserting `0` which would be a no-op

Reply:
Done.